mantis/simulation/routers/solver.py

```diff
@@ -0,0 +1,205 @@
+                                    fee_out=0
+                    )
+            )
+        return exchanges, assets
+
+    def get_new_reserve(
+        self, delta: Variable, lamb: Variable, exchange: RouterExchange
+    ) -> Expression:
+        return exchange.reserves + (1 - exchange.fee_in) * delta - (1 - exchange.fee_out) * lamb
+
+    def solve(self, data: AllData, input: Input) -> Output:
+        if not input.max:
+            raise NotImplementedError("'max' value on input is not supported to be False yet")
+        exchanges, assets = self.parse_data(data)
+        return self._solve(exchanges, assets, input)
+
+    def _solve(
+        self, exchanges: list[RouterExchange], assets: list[RouterAsset], input: Input
+    ) -> Output:
+        def asset_index(token_id: int) -> int:
+            return [i for i, asset in enumerate(assets) if asset.id == token_id][0]
+        
+        is_micp = len(exchanges) <= self.MICP_THRESHOLD
+        num_exchanges = len(exchanges)
+        num_assets = len(assets)
+
+        initial_assets = np.zeros(num_assets)  # Initial assets
+
+        initial_assets[asset_index(input.in_token_id)] = input.in_amount
+
+        A = []
+        for exchange in exchanges:
+            n_i = len(exchange.assets)  # Number of tokens in pool (default to 2)
+            A_i = np.zeros((num_assets, n_i))
+            for i, token in enumerate(exchange.assets):
+                A_i[asset_index(token.id), i] = 1
+            A.append(A_i)
+
+        # tendered (given) amount
+        deltas = [
+            cp.Variable(len(exchange.assets), nonneg=True) for exchange in exchanges
+        ]
+
+        # received (wanted) amounts
+        lambdas = [
+            cp.Variable(len(exchange.assets), nonneg=True) for exchange in exchanges
+        ]
+        
+        # Binary value, indicates tx or not for given pool
+        eta = cp.Variable(
+            num_exchanges, boolean=is_micp, nonneg=not is_micp
+        )
+
+        # network trade vector - net amount received over all trades(transfers/exchanges)
+        psi = cp.sum(
+            [A_i @ (LAMBDA - DELTA) for A_i, DELTA, LAMBDA in zip(A, deltas, lambdas)]
+        )
+
+        # Objective is to trade number_of_init_tokens of asset origin_token for a maximum amount of asset objective_token
+        obj = cp.Maximize(
+            psi[asset_index(input.out_token_id)]
+            - cp.sum(eta @ [exchange.fix_cost for exchange in exchanges])
+        )
+
+        # Trading function constraints
+        constrains = [
+            psi + initial_assets >= 0,
+        ]
+
+        for i, exchange in enumerate(exchanges):
+            new_reserve = self.get_new_reserve(deltas[i], lambdas[i], exchange)
+            constrains.extend(exchange.constraints(new_reserve))
+            if not is_micp:
+                constrains.append(eta >= 0)
+                constrains.append(eta <= 15)
+                constrains.append(
```

Comment:
hard to follow number without commnets/tests

Reply:
@raserhin if you will have some time to explain why rules exists as is, could be awesome.